### PR TITLE
[[ BrowserWidget ]] Implementation of a web browser as a widget

### PR DESCRIPTION
**Pull request by livecodeian**
[Review on Reviewable](https://reviewable.io/reviews/runrev/livecode/2769)

### CODE REVIEW

**Comment by peter-b** [comment]
Hi @runrevian -- is there a spec for this that I should be reviewing against?

**Comment by peter-b** [comment]
Review status: 0 of 90 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

<sup>**[engine/src/browser.lcb, line 2 [r1]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxeZBfxhRieg-gfan35)** ([raw file](https://github.com/runrev/livecode/blob/6e1b6d4fb4f50aab6ab71d5fc7ae627224fef622/engine/src/browser.lcb#L2)):</sup>
I think this should be LiveCode Ltd.!

---

<sup>**[engine/src/browser.lcb, line 86 [r1]](https://reviewable.io:443/reviews/runrev/livecode/2769#-Jxe_MwDWpUTtoYyEbeK)** ([raw file](https://github.com/runrev/livecode/blob/6e1b6d4fb4f50aab6ab71d5fc7ae627224fef622/engine/src/browser.lcb#L86)):</sup>
I think `pSize` should be `rSize` here.

Strictly speaking, `uint32_t` translates to `UInt32` in LCB.  `CUInt` is C's `unsigned int` type.

This applies to other foreign handler declarations in this PR.

---

<sup>**[engine/src/browser.lcb, line 116 [r1]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxeZp-Hek0semVP8qDL)** ([raw file](https://github.com/runrev/livecode/blob/6e1b6d4fb4f50aab6ab71d5fc7ae627224fef622/engine/src/browser.lcb#L116)):</sup>
It would be nice to be able to write:

```
kMCBrowserPropertyMap[kMCBrowserPropertyScrollbars]
```

Since LCB lists are 1-indexed, why not number the constants from 1 rather than from 0? It might simplify `browserListToLCBList`, for example. This applies to the other enumeration-like declarations in this PR.

---

<sup>**[engine/src/cmds.cpp, line 516 [r1]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxebLj_4jhP7dKaJJkg)** ([raw file](https://github.com/runrev/livecode/blob/6e1b6d4fb4f50aab6ab71d5fc7ae627224fef622/engine/src/cmds.cpp#L516)):</sup>
Do you need to check that `t_object` is actually an `MCWidget` before making this call?  As I understand it, `MCChunk->getobj()` will happily evaluate to any object instance.

---

<sup>**[engine/src/java/com/runrev/android/libraries/LibBrowser.java, line 51 [r1]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxecMA1w-Y0Ya_SKG1s)** ([raw file](https://github.com/runrev/livecode/blob/6e1b6d4fb4f50aab6ab71d5fc7ae627224fef622/engine/src/java/com/runrev/android/libraries/LibBrowser.java#L51)):</sup>
Please remove this commented-out code, or add a comment to the source code explaining why it should be kept! (same applies elsewhere in this file)

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/runrev/livecode/2769)

**Comment by peter-b** [comment]
Reviewed 9 of 90 files at r1.
Review status: 9 of 90 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/runrev/livecode/2769)

**Comment by livecodeian** [comment]
Review status: 9 of 90 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

<sup>**[engine/src/browser.lcb, line 2 [r1]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxeZBfxhRieg-gfan35)** ([raw file](https://github.com/runrev/livecode/blob/6e1b6d4fb4f50aab6ab71d5fc7ae627224fef622/engine/src/browser.lcb#L2)):</sup>
Yeah, I'll update those in any new files I've introduced, though there's a lot of references to Runtime Revolution still in the develop branch.

---

<sup>**[engine/src/browser.lcb, line 86 [r1]](https://reviewable.io:443/reviews/runrev/livecode/2769#-Jxe_MwDWpUTtoYyEbeK)** ([raw file](https://github.com/runrev/livecode/blob/6e1b6d4fb4f50aab6ab71d5fc7ae627224fef622/engine/src/browser.lcb#L86)):</sup>
Yup, will change that.

---

<sup>**[engine/src/browser.lcb, line 116 [r1]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxeZp-Hek0semVP8qDL)** ([raw file](https://github.com/runrev/livecode/blob/6e1b6d4fb4f50aab6ab71d5fc7ae627224fef622/engine/src/browser.lcb#L116)):</sup>
The enumeration values are taken from the C API, which appropriately enough is 0-based. Not going to change this.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/runrev/livecode/2769)

**Comment by livecodeian** [comment]
Review status: 9 of 90 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[engine/src/cmds.cpp, line 516 [r1]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxebLj_4jhP7dKaJJkg)** ([raw file](https://github.com/runrev/livecode/blob/6e1b6d4fb4f50aab6ab71d5fc7ae627224fef622/engine/src/cmds.cpp#L516)):</sup>
Yup, should be performing the same typecheck as the other syntax I added here.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/runrev/livecode/2769)

**Comment by peter-b** [comment]
Reviewed 6 of 90 files at r1, 1 of 27 files at r2.
Review status: 16 of 90 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

<sup>**[engine/src/module-engine.cpp, line 747 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxfaXBLGmlUsKUqnOF5)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/engine/src/module-engine.cpp#L747)):</sup>
Hi Ian,

Thanks for addressing my more minor quibbles from earlier this evening so promptly! :+1: 

Unfortunately, we might have a bit of a problem here.

On the Emscripten platform, any code path that can (eventually) call `MCUIDC::wait()` can't be executed directly -- every function in any of its call paths has to be compiled to (another) intermediate bytecode and interpreted (using a mechanism called Emterpreter).

As it currently stands, LiveCode Builder is rather slow.  If LiveCode Builder code can enter an inner main loop (via `MCEngineRunloopWait()`, specifically), most of libscript -- and particularly the performance-critical handler dispatch code -- will have to be compiled with Emterpreter.  Which means we'll have an VM written in C++ (e.g. SpiderMonkey) running a Emterpreter bytcode interpreter running a LiveCode Builder bytecode interpreter running LCB. Which will be **really** slow, even in the best of all possible circumstances. :sob: It'll also preclude ever compiling LiveCode Builder code directly to JavaScript (which would let us take advantage of browsers' JIT JavaScript compilation capabilities).

It would be really, really helpful to our widgets-in-HTML5 plans if you could find a way to refactor the browser widget branch in such a way that `MCEngineRunloopWait()` can be eliminated! There's no problem with `AddRunloopAction()`, `RemoveRunloopAction()` -- we can even figure out some way to add one-shot callbacks to the top-level main loop, if necessary.  I think that HTML5 is going to be a really important target platform and revenue stream, so making sure that we don't take a massive hit to performance right from the outset seems like a priority to me.

I'd be really happy to organise a Skype call with some people (@livecodefraser, @livecodeali  along with the two of us, maybe?) if we need to talk through this and find a way to solve the problems...! Let me know. Sorry to be the bearer of bad news... :disappointed: 

Thanks again!

Peter

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/runrev/livecode/2769)

**Comment by peter-b** [comment]
Review status: 16 of 90 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[engine/src/browser.lcb, line 86 [r1]](https://reviewable.io:443/reviews/runrev/livecode/2769#-Jxe_MwDWpUTtoYyEbeK)** ([raw file](https://github.com/runrev/livecode/blob/6e1b6d4fb4f50aab6ab71d5fc7ae627224fef622/engine/src/browser.lcb#L86)):</sup>
Are you planning to update the types here as well?

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/runrev/livecode/2769)

**Comment by livecodeian** [comment]
<sup>**[engine/src/browser.lcb, line 86 [r1]](https://reviewable.io:443/reviews/runrev/livecode/2769#-Jxe_MwDWpUTtoYyEbeK)** ([raw file](https://github.com/runrev/livecode/blob/6e1b6d4fb4f50aab6ab71d5fc7ae627224fef622/engine/src/browser.lcb#L86)):</sup>
in foreign.mlc: public type CUInt is UInt32
So currently they're the same.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/runrev/livecode/2769)

File: libbrowser/libbrowser.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -0,0 +1,432 @@
+// !$*UTF8*$!
```
Comment (livecodefraser) [inline libbrowser/libbrowser.xcodeproj/project.pbxproj]: The project file should be generated by Gyp and shouldn't be part of the source tree.

**Comment by peter-b** [comment]
Reviewed 23 of 90 files at r1, 3 of 27 files at r2.
Review status: 42 of 90 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

<sup>**[engine/src/exec-interface-widget.cpp, line 45 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxicQJnyCGvtdRiAk_a)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/engine/src/exec-interface-widget.cpp#L45)):</sup>
Please check this (in general, please don't leave anything `/* UNCHECKED */` in new code).

---

<sup>**[engine/src/native-layer-mac.mm, line 195 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxiXNR4gcQJ_HYhYome)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/engine/src/native-layer-mac.mm#L195)):</sup>
It's not obvious to me why it makes sense to pass a zero-sized rectangle here.  Could you please add a comment with a brief explanation?

Is the idea to call `OnGeometryChanged()` for its side-effects?

---

<sup>**[engine/src/native-layer-win32.cpp, line 116 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxiXwrJzqLD3le9N4kV)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/engine/src/native-layer-win32.cpp#L116)):</sup>
`OnGeometryChanged()` with invalid rectangle.

---

<sup>**[engine/src/native-layer-x11.cpp, line 222 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxiY_6b5jIu2n5DZ5iV)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/engine/src/native-layer-x11.cpp#L222)):</sup>
`OnGeometryChanged()` with invalid rectangle.

---

<sup>**[engine/src/uidc.cpp, line 948 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxiZNjv25sXu6RXxLdb)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/engine/src/uidc.cpp#L948)):</sup>
Why not add `MCRunloopAction::Retain()` and `MCRunloopAction::Release()` methods?  It would mirror the other reference-counting APIs we currently use.

---

<sup>**[engine/src/widget.cpp, line 888 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-Jxi_wqQ-s4u4X7uo7yQ)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/engine/src/widget.cpp#L888)):</sup>
Could you please factor this out so that you have only one assignment per line? Something like:

``` C++
if (t_success && (p_view != nil))
{
    t_layer = MCNativeLayer::CreateNativeLayer(getwidget(), p_view);
    t_success = (nil != t_layer);
}
```

It's only one extra line of code but it improves the readability quite a lot.

---

<sup>**[engine/src/widget.mlc, line 1496 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxibEU42rPET0LeP7NS)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/engine/src/widget.mlc#L1496)):</sup>
Please add docs for `native layer of _` and `my native layer`.

---

<sup>**[extensions/widgets/browser/browser.lcb, line 56 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-Jxibsefqfaohz62TPWW)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/extensions/widgets/browser/browser.lcb#L56)):</sup>
Please add documentation comment blocks for these properties.

---

<sup>**[extensions/widgets/browser/browser.lcb, line 352 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxicgMZY2VmpikCeeQR)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/extensions/widgets/browser/browser.lcb#L352)):</sup>
Please always check results in LCB code (even if you just immediately `throw`).

---

<sup>**[libbrowser/include/libbrowser.h, line 103 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxieBwf0IZy68h8pbgJ)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/libbrowser/include/libbrowser.h#L103)):</sup>
These actions could take a long time to complete.  Is this API supposed to be synchronous or asynchronous (from the prototypes, I guess that they block)?

---

<sup>**[libbrowser/libbrowser.xcodeproj/project.pbxproj, line 1 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-Jxif6Ge_-DsAkVmA6ag)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/libbrowser/libbrowser.xcodeproj/project.pbxproj#L1)):</sup>
I think you should probably remove this file, since gyp generates its own xcodeproj files.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/runrev/livecode/2769)

**Comment by livecodeian** [comment]
Review status: 41 of 91 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

<sup>**[engine/src/module-engine.cpp, line 747 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxfaXBLGmlUsKUqnOF5)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/engine/src/module-engine.cpp#L747)):</sup>
Hi Peter. I understand the concern here, however I'm not sure that this is actually a practical problem for libbrowser. I can't see us bundling the entirety of libcef into an html5 standalone (it's just too big) so it's probable that another means of implementing a browser on that platform will be needed - similarly for the Android browser implementation where MCEngineRunloopWait is also used.

How we implement a browser widget for use with HTML5 is I think beyond the scope of this pull request.

---

<sup>**[engine/src/native-layer-mac.mm, line 195 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxiXNR4gcQJ_HYhYome)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/engine/src/native-layer-mac.mm#L195)):</sup>
OnGeometryChanged is inconsistent with other On*Changed methods in that the passed value represents the old rectangle rather than the new. I'm not sure why this was ever the case, as in each implementation the old value is ignored anyway. I'd be in favour of removing that parameter entirely, or replacing it with the new geometry value instead. A question for Fraser to address perhaps?

---

<sup>**[engine/src/uidc.cpp, line 948 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxiZNjv25sXu6RXxLdb)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/engine/src/uidc.cpp#L948)):</sup>
I only intended the reference counting here to be internal housekeeping. If you can provide a legitimate use-case for keeping multiple references to a runloop action then I'll consider adding that in.

---

<sup>**[libbrowser/include/libbrowser.h, line 103 [r2]](https://reviewable.io:443/reviews/runrev/livecode/2769#-JxieBwf0IZy68h8pbgJ)** ([raw file](https://github.com/runrev/livecode/blob/81364ece0f66385c568accc8dee86467c2722b9c/libbrowser/include/libbrowser.h#L103)):</sup>
Browser navigation is asynchronous, navigation & load events are signalled through the callback function. EvaluateJavaScript is synchronous and returns the resulting value of the javascript call.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/runrev/livecode/2769)

**Comment by livecodefraser** [comment]
Reviewed 6 of 90 files at r1, 2 of 4 files at r4, 1 of 3 files at r5.
Review status: 50 of 91 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/runrev/livecode/2769)

**Comment by livecodefraser** [comment]
Reviewed 7 of 90 files at r1, 1 of 3 files at r5.
Review status: 58 of 91 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/runrev/livecode/2769)

**Comment by livecode-vulcan** [comment]
📝 Hi @runrevian, I haven't been able to verify that you've signed our [LiveCode Contributor's Agreement](https://livecode.com/account/developer/contribute).

If you have previously signed the Contributor's Agreement, I may not be able to detect it because you haven't [linked your GitHub account to your LiveCode account](https://livecode.com/account/developer/github).

Please see the [information for contributors](https://github.com/runrev/livecode/blob/develop/HACKING.md) for more information.

**Comment by livecode-vulcan** [comment]
📝 Hi @runrevian, I haven't been able to verify that you've signed our [LiveCode Contributor's Agreement](https://livecode.com/account/developer/contribute).

If you have previously signed the Contributor's Agreement, I may not be able to detect it because you haven't [linked your GitHub account to your LiveCode account](https://livecode.com/account/developer/github).

Please see the [information for contributors](https://github.com/runrev/livecode/blob/develop/HACKING.md) for more information.

**Comment by livecodefraser** [comment]
Reviewed 85 of 89 files at r1.
Review status: 85 of 89 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/2769)

**Comment by livecodefraser** [comment]
Reviewed 4 of 89 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/2769)

**Comment by livecodefraser** [comment]
@livecode-vulcan review okay 73106ad

**Comment by livecode-vulcan** [comment]
💙 review by @livecodefraser ok 73106adf0b1819b1ccd585af105c60d233612cc4

**Comment by livecode-vulcan** [comment]
😎 test success 73106adf0b1819b1ccd585af105c60d233612cc4
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/693)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/706)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/694)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/696)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/706)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/717)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/708)

**Comment by livecodefraser** [comment]
@livecode-vulcan review okay f24a439

**Comment by livecode-vulcan** [comment]
💙 review by @livecodefraser ok f24a439f275af02307bbb7b680f849ec39926ce8

**Comment by livecode-vulcan** [comment]
😎 test success f24a439f275af02307bbb7b680f849ec39926ce8
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/727)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/740)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/728)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/730)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/740)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/751)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/742)

**Pull request merged**